### PR TITLE
rpc: show P2(W)SH redeemScript in getrawtransaction #27637

### PR DESCRIPTION
**Pull request by Riahiamirreza**
This a work in progress PR. I'll squash all my commits at the end to make commits atomic (as mentioned [here](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches))
It's an effort to solve this issue https://github.com/bitcoin/bitcoin/issues/27391.

### CODE REVIEW

**Comment by pinheadmz** [comment]
@Riahiamirreza are you still working on this?

**Comment by Riahiamirreza** [comment]
@pinheadmz Yes, but I have some difficulties in creating a transaction which spends from a P2SH to be able to view the structure of the transaction.

**Comment by Riahiamirreza** [comment]
@MarcoFalke Actually in the last couple of weeks I didn't put effort on it. If you think this is not critical and can be delayed more, I need more time to spend on it. I know it's somewhat a basic issue in the Bitcoin core, but it's my first issue.

**Comment by Riahiamirreza** [comment]
I'm fixing tests.

File: src/core_write.cpp [hunk 1]
```diff
@@ -202,6 +202,19 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
...
+            CScript::const_iterator pc = txin.scriptSig.begin();
+            txin.scriptSig.GetOp(pc, opcode, vch);
+            txin.scriptSig.GetOp(pc, opcode, vch);
+            txin.scriptSig.GetOp(pc, opcode, vch);
+            txin.scriptSig.GetOp(pc, opcode, vch);
+            CScript script;
+            std::vector<unsigned char> scriptData(ParseHex(HexStr(vch)));
+            script = CScript(scriptData.begin(), scriptData.end());
+            ScriptToUniv(script, /*out=*/redeemScript, /*include_hex=*/false, /*include_address=*/true);
+            in.pushKV("redeemScrtipt", redeemScript);
```
Comment (Sjors) [inline src/core_write.cpp:217]: `redeemScrtipt` is a typo

**Comment by Sjors** [comment]
Fixing the test is hopefully a matter of just adding the new field(s) to these tests.

Also make sure to [squash your commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits).

**Comment by Sjors** [comment]
> I think it's not a convenient way to call `txin.scriptSig.GetOp(pc, opcode, vch);` three times.

Indeed it's not. But I don't understand what you're trying to do here.

I would suggest first fixing the tests, so that it's clear if the code is correct. And then worry about making the code nicer.

**Comment by Riahiamirreza** [comment]
@Sjors For adding new field to a transaction (in my case, adding `redeemScript` to `vin`) is it correct to add new fields to `CTxIn` class in `test/functional/test_framework/messages.py`? While adding it will cause mismatch in weight calculation of transactions, I think it's not the correct way. But I'm not sure what's the correct way.

**Comment by Riahiamirreza** [comment]
@sipa Yes and I'm not sure how should I fix the problem. Actually the `redeemScript` is a new field which is not part of the actual data, it's only a new representation of data (decompiled version of `redeemScript`). Would you suggest any approach to fix it?

**Comment by Riahiamirreza** [comment]
@sipa @maflcko I added new field to the output of `TxToUniv` function, called `redeemScript`. I've manually tested the rpc and that works fine. After running tests I faced issue of decoding transactions. Because of adding a new field, transactions couldn't be decoded successfully. I attempt to change `CTxIn` message in my next effort and add the `redeemScript` which now seems trivially incorrect way.

**Comment by Riahiamirreza** [comment]
@maflcko Sure, I'll revert my changes on test framework and then try another way to fix tests.

**Comment by Riahiamirreza** [comment]
```feature_abortnode.py                                   | ✖ Failed  | 1 s
feature_bip68_sequence.py                              | ✖ Failed  | 1 s
feature_cltv.py                                        | ✖ Failed  | 1 s
feature_coinstatsindex.py                              | ✖ Failed  | 2 s
feature_config_args.py                                 | ✖ Failed  | 10 s
feature_csv_activation.py                              | ✖ Failed  | 1 s
feature_dersig.py                                      | ✖ Failed  | 1 s
feature_fee_estimation.py                              | ✖ Failed  | 1 s
feature_init.py                                        | ✖ Failed  | 63 s
feature_nulldummy.py                                   | ✖ Failed  | 1 s
feature_rbf.py                                         | ✖ Failed  | 1 s
feature_utxo_set_hash.py                               | ✖ Failed  | 1 s
interface_rest.py                                      | ✖ Failed  | 1 s
mempool_accept.py                                      | ✖ Failed  | 1 s
mempool_datacarrier.py                                 | ✖ Failed  | 1 s
mempool_dust.py                                        | ✖ Failed  | 1 s
mempool_expiry.py                                      | ✖ Failed  | 1 s
mempool_limit.py                                       | ✖ Failed  | 1 s
mempool_package_limits.py                              | ✖ Failed  | 1 s
mempool_package_onemore.py                             | ✖ Failed  | 1 s
mempool_packages.py                                    | ✖ Failed  | 1 s
mempool_persist.py --descriptors                       | ✖ Failed  | 1 s
mempool_reorg.py                                       | ✖ Failed  | 1 s
mempool_resurrect.py                                   | ✖ Failed  | 1 s
mempool_sigoplimit.py                                  | ✖ Failed  | 1 s
mempool_spend_coinbase.py                              | ✖ Failed  | 1 s
mempool_unbroadcast.py                                 | ✖ Failed  | 7 s
mempool_updatefromblock.py                             | ✖ Failed  | 1 s
mining_basic.py                                        | ✖ Failed  | 12 s
mining_getblocktemplate_longpoll.py                    | ✖ Failed  | 8 s
mining_prioritisetransaction.py                        | ✖ Failed  | 1 s
p2p_blockfilters.py                                    | ✖ Failed  | 491 s
p2p_compactblocks.py                                   | ✖ Failed  | 2 s
p2p_feefilter.py                                       | ✖ Failed  | 2 s
p2p_filter.py                                          | ✖ Failed  | 1 s
p2p_ibd_txrelay.py                                     | ✖ Failed  | 1 s
p2p_invalid_messages.py                                | ✖ Failed  | 4 s
p2p_leak_tx.py                                         | ✖ Failed  | 1 s
p2p_permissions.py                                     | ✖ Failed  | 8 s
p2p_segwit.py                                          | ✖ Failed  | 4 s
rpc_blockchain.py                                      | ✖ Failed  | 14 s
rpc_createmultisig.py                                  | ✖ Failed  | 1 s
rpc_decodescript.py                                    | ✖ Failed  | 1 s
rpc_deriveaddresses.py                                 | ✖ Failed  | 1 s
rpc_deriveaddresses.py --usecli                        | ✖ Failed  | 1 s
rpc_dumptxoutset.py                                    | ✖ Failed  | 1 s
rpc_generate.py                                        | ✖ Failed  | 1 s
rpc_invalid_address_message.py                         | ✖ Failed  | 1 s
rpc_mempool_info.py                                    | ✖ Failed  | 1 s
rpc_misc.py                                            | ✖ Failed  | 1 s
rpc_net.py                                             | ✖ Failed  | 1 s
rpc_packages.py                                        | ✖ Failed  | 1 s
rpc_rawtransaction.py --legacy-wallet                  | ✖ Failed  | 1 s
rpc_scanblocks.py                                      | ✖ Failed  | 1 s
rpc_scantxoutset.py                                    | ✖ Failed  | 1 s
rpc_txoutproof.py                                      | ✖ Failed  | 1 s

ALL                                                    | ✖ Failed  | 1165 s (accumulated)
```

These are the currently failing tests.

**Comment by Riahiamirreza** [comment]
> Failing why? Without any information, there is nothing we can do here.
As the exception suggests it's due the addition of the new field `redeemScript`.
Am I right?
```
{a.riahi@bitcoin} > python3.10 ./test/functional/rpc_generate.py
2023-11-13T14:57:56.863000Z TestFramework (INFO): PRNG seed is: 2176820945159495690
2023-11-13T14:57:56.863000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_nt1lv9on
2023-11-13T14:57:57.137000Z TestFramework (INFO): Test rpc generate raises with message to use cli option
2023-11-13T14:57:57.138000Z TestFramework (INFO): Test rpc generate help prints message to use cli option
2023-11-13T14:57:57.138000Z TestFramework (INFO): Test rpc generate is a hidden command not discoverable in general help
2023-11-13T14:57:57.234000Z TestFramework (INFO): Mine an empty block to address and return the hex
2023-11-13T14:57:57.237000Z TestFramework (INFO): Generate an empty block to address
2023-11-13T14:57:57.241000Z TestFramework (INFO): Generate an empty block to a descriptor
2023-11-13T14:57:57.244000Z TestFramework (INFO): Generate an empty block to a combo descriptor with compressed pubkey
2023-11-13T14:57:57.247000Z TestFramework (INFO): Generate an empty block to a combo descriptor with uncompressed pubkey
2023-11-13T14:57:57.251000Z TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/home/a.riahi@asax.local/github_repos/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/home/a.riahi@asax.local/github_repos/bitcoin/./test/functional/rpc_generate.py", line 22, in run_test
    self.test_generateblock()
  File "/home/a.riahi@asax.local/github_repos/bitcoin/./test/functional/rpc_generate.py", line 68, in test_generateblock
    miniwallet.send_self_transfer(from_node=node)
  File "/home/a.riahi@asax.local/github_repos/bitcoin/test/functional/test_framework/wallet.py", line 247, in send_self_transfer
    self.sendrawtransaction(from_node=from_node, tx_hex=tx['hex'])
  File "/home/a.riahi@asax.local/github_repos/bitcoin/test/functional/test_framework/wallet.py", line 355, in sendrawtransaction
    self.scan_tx(from_node.decoderawtransaction(tx_hex))
  File "/home/a.riahi@asax.local/github_repos/bitcoin/test/functional/test_framework/coverage.py", line 49, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/a.riahi@asax.local/github_repos/bitcoin/test/functional/test_framework/authproxy.py", line 126, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: Internal bug detected: RPC call "decoderawtransaction" returned incorrect type:
{
    "vin": {
        "0": {
            "redeemScript": "key returned that was not in doc"
        }
    }
}
Bitcoin Core v26.99.0-8f567b65a345-dirty
Please report this issue here: https://github.com/bitcoin/bitcoin/issues
 (-1)

```

File: src/core_write.cpp [hunk 2]
```diff
@@ -202,6 +202,21 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
...
             in.pushKV("scriptSig", o);
+            if (txin.scriptSig.IsPayToScriptHash()) {
+                UniValue redeemScript(UniValue::VOBJ);
+                opcodetype opcode;
+                std::vector<unsigned char> vch;
+                CScript::const_iterator pc = txin.scriptSig.begin();
+                txin.scriptSig.GetOp(pc, opcode, vch);
+                txin.scriptSig.GetOp(pc, opcode, vch);
+                txin.scriptSig.GetOp(pc, opcode, vch);
+                txin.scriptSig.GetOp(pc, opcode, vch);
```
[the remark this answers is not in the document]
Comment (Riahiamirreza) [reply src/core_write.cpp:213]: Is there any convenient way to detect redeemScript in the script?

[the remark this answers is not in the document]
Comment (Riahiamirreza) [reply src/core_write.cpp:213]: So can I always expect the last `scriptSig` push to be the `redeemScript` if it has one?

**Comment by Riahiamirreza** [comment]
@sipa As far as I understand having the information of UTXOs has nothing to do with pruned mode. Don't pruned nodes store the whole UTXO set? (I can guess the problem can occur for pruned mode when a reorg happen and the old data to restore is pruned, but I'm not sure this is the case)

File: src/core_write.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Riahiamirreza) [reply src/core_write.cpp:213]: @sipa Do you think this is a proper way of accessing the last push of the `scriptSig`? Or there are utilities in the codebase to do this for me?
```c++
            while (true) {
                if (txin.scriptSig.GetOp(pc, opcode, vch))
                    break;
            }
```

File: src/core_write.cpp [hunk 3]
```diff
@@ -202,6 +202,21 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
             o.pushKV("asm", ScriptToAsmStr(txin.scriptSig, true));
             o.pushKV("hex", HexStr(txin.scriptSig));
             in.pushKV("scriptSig", o);
+            if (txin.scriptSig.IsPayToScriptHash()) {
```
[the remark this answers is not in the document]
Comment (Riahiamirreza) [reply src/core_write.cpp:205]: @sipa How can I access the `CTxout` from the `COutPoint`? For a given `CTxIn` I want to reach the `scriptPubKey` of its output. The `COutPoint` only has transaction hash and index of that specific output in the `vout`. I don't know how can I reach a transaction output using `COutPoint`. Should I use `CCoinsViewDB::GetCoin`? Does it contains all the transactions? (not only mempool transactions)
```cpp
    bool GetCoin(const COutPoint &outpoint, Coin &coin) const override;
```

[the remark this answers is not in the document]
Comment (Riahiamirreza) [reply src/core_write.cpp:205]: @sipa The undo data contains Coins, which only have amount and scriptPubKey. The Txins' outpoint have tx hash and index of vout. My problem is how to identify which one of Coins in the undo data match to a given txin's prevout.

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -114,6 +114,12 @@ static std::vector<RPCResult> DecodeTxDoc(const std::string& txid_field_doc)
                     {RPCResult::Type::STR, "asm", "Disassembly of the signature script"},
                     {RPCResult::Type::STR_HEX, "hex", "The raw signature script bytes, hex-encoded"},
                 }},
+                {RPCResult::Type::OBJ, "redeemScript", /*optional=*/true, "",
```
Comment (Riahiamirreza) [inline src/rpc/rawtransaction.cpp:117]: I'll add description to description fields later.

**Comment by Riahiamirreza** [comment]
I'm working on adding tests, but I'm not sure where is the correct place for putting tests. Should I put tests in `RawTransactionsTest.getrawtransaction_verbositry_tests` in the `tests/functional/rpc_rawtransaction.py`? I need to have a transaction of type P2SH in the chain and check whether the result has the correct `redeemScript` field? Is this approach correct?

**Comment by Riahiamirreza** [comment]
@sipa Would you please review my PR? I really appreciate it.

**Comment by Riahiamirreza** [comment]
@maflcko Would you please review the PR if it worth? If you think it does not have enough quality or is out of priority just ignore my comment.

**Comment by Riahiamirreza** [comment]
@maflcko You're right, I didn't put much effort on it. Do you think the issue is still relevant?

**Pull request closed without merging**